File: web-examples/php/sendVote.php
<?php
include_once('Crypt/RSA.php');

$domain = '';

#
# This is example code on how to send a vote to Votifier-mk2
# Votes cosist of a json encoded hash with the values
# Type, serviceName, username, address, timestamp, uuid
# All extra values will be ignored by Votifier-mk2
#
# The Json encoded hash (Our vote) is then signed
# The signature is then appended to the end of the vote string with '\xa7' as a delimiter
# vote + '\xa7' + signature
# And finally sent to the server
#

function load_keys($ident, $type) {
	if ($type == 'public') {
		$key = file_get_contents($ident.'.pub');
	} elseif ($type == 'private') {
		$key = file_get_contents($ident);
	}
	return $key;
}

function rsa_sign($key, $message) {
	$rsa = new Crypt_RSA(); 
	$rsa->setHash("sha256"); 
	$rsa->setSignatureMode(CRYPT_RSA_SIGNATURE_PKCS1); 
	$rsa->loadKey($key); 
	$signature = $rsa->sign($message);
	$signature = base64_encode($signature);
	return $signature;
}

function build_vote($user, $addr, $domain) {
	$epoch = date('U');
	$uuid = hash('sha1', $epoch.$domain.$user.$addr);
	$vote = json_encode(array("Type"=>"VOTE", "serviceName"=>$domain, "username"=>$user, "address"=>$addr, "timestamp"=>$epoch, "uuid"=>$uuid));
	return $vote;
}

function build_payload($user, $addr, $domain) {
	$vote = build_vote($user, $addr, $domain);
	$sign = rsa_sign(load_keys('id_servlist', 'private'), $vote);
	$payload = $vote . '\xa7' . $sign;
	return $payload;
}

function send_vote($server, $port, $user, $addr, $domain) {
	$payload = build_payload($user, $addr, $domain);
	$socket = socket_create(AF_INET, SOCK_STREAM, SOL_TCP);
	if ($socket === false) {
		return "socket_create() failed: reason: " . socket_strerror(socket_last_error()) . "\n";
	}
	$result = socket_connect($socket, $addr, (int) $port);
	if ($result === false) {
		return "socket_connect() failed.\nReason: ($result) " . socket_strerror(socket_last_error($socket)) . "\n";
	}
	socket_write($socket, $payload);
	socket_read($socket, 2048);
	socket_close($socket);
}

$output = send_vote('127.0.0.1', '35566', 'KsaRedFx', '192.168.0.1', $domain);
print $output;


?>

File: web-examples/python/sendVote.py
import simplejson as json
import socket
from Crypto.PublicKey import RSA 
from Crypto.Signature import PKCS1_v1_5 
from Crypto.Cipher import PKCS1_OAEP 
from Crypto.Hash import SHA, SHA256
import base64
import time
import os

#
# This is example code on how to send a vote to Votifier-mk2
# Votes cosist of a json encoded hash with the values
# Type, serviceName, username, address, timestamp, uuid
# All extra values will be ignored by Votifier-mk2
#
# The Json encoded hash (Our vote) is then signed
# The signature is then appended to the end of the vote string with '\xa7' as a delimiter
# vote + '\xa7' + signature
# And finally sent to the server
#

domain = ''
keyname = 'id_servlist'

def signVote(private, message):
  """ Generates a signature for the vote message """
  rsakey = RSA.importKey(private) 
  signer = PKCS1_v1_5.new(rsakey) 
  digest = SHA256.new() 
  digest.update(message) 
  sign = base64.b64encode(signer.sign(digest))
  return sign

def generateVote(user, addr):
  """ Creates the vote off of information given """
  epoch = time.time()
  uuid = SHA.new("%s%s%s%s%s" % (epoch,domain,os.urandom(25),user,addr)).hexdigest() # Generates a random UUID - This can be anything
  return json.dumps(dict(Type="VOTE", serviceName=domain, username=user, address=addr, timestamp=epoch, uuid=uuid))

def prepareVote(user, addr):
  """ Calls Generate vote, opens your keyfile, 
      and then signs the vote with they key 
  """
  vote = generateVote(user, addr)
  f = open(keyname, 'r').read()
  sign = signVote(f, vote)
  message = vote + '\xa7' + sign
  return message

def sendVote(server, port, user, addr):
  """ Sends the vote off to the server with a bit of redundancy for blocked ports """
  message = prepareVote(user, addr)
  s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
  s.settimeout(8)
  try:
    s.connect((server, int(port)))
  except:
    try:
      s.close()
    except:
      pass
    s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    s.settimeout(8)
    s.connect((server, int(port)))
  s.send(message)


sendVote(server='127.0.0.1', port='35566', user='KsaRedFx', addr='192.168.0.1')